cpp/unreal_plugins/SpEngine/Source/SpEngine/WorkQueue.h

```diff
@@ -0,0 +1,125 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#pragma once
+
+#include <concepts>    // std::same_as
+#include <future>
+#include <map>
+#include <mutex>
+#include <string>
+#include <utility>     // std::forward, std::move
+#include <vector>
+#include <type_traits> // std::invoke_result_t
+
+#include <SpCore/Assert.h>
+#include <SpCore/Boost.h>
+
+template <typename TFunc>
+concept CFuncIsCallableWithNoArgs = requires(TFunc func) {
+    { func() };
+};
+
+template <typename TFunc, typename... TArgs>
+concept CFuncIsCallableWithArgs = requires(TFunc func, TArgs... args) {
+    { func(args...) };
+};
+
+template <typename TFunc, typename TReturn, typename... TArgs>
+concept CFuncReturnsAndIsCallableWithArgs = requires(TFunc func, TArgs... args) {
+    { func(args...) } -> std::same_as<TReturn>;
+};
+
+class WorkQueue {
+
+public:
+    WorkQueue() : io_context_(), executor_work_guard_(io_context_.get_executor()) {}
+
+    void run()
+    {
+        // run all scheduled work and wait for executor_work_guard_.reset() to be called from a worker thread
+        io_context_.run();
+
+        // reinitialize io_context_ and executor_work_guard_ to prepare for the next call to run()
+        mutex_.lock();
+        io_context_.restart();
+        new(&executor_work_guard_) boost::asio::executor_work_guard<boost::asio::io_context::executor_type>(io_context_.get_executor());
+        mutex_.unlock();
+    }
+
+    void reset()
+    {
+        // request io_context_.run() to stop executing once all of its scheduled work is finished
+        mutex_.lock();
+        executor_work_guard_.reset();
+        mutex_.unlock();
+    }
+
+    template <typename TFunc>
+    static auto wrapFuncToExecuteInWorkQueueBlocking(WorkQueue& work_queue, TFunc&& func)
+    {
+        return wrapFuncToExecuteInWorkQueueBlockingImpl(work_queue, std::forward<TFunc>(func), FuncInfo<TFunc>());
+    }
+
+private:
+
+    template <typename TClass>
+    struct FuncInfo : public FuncInfo<decltype(&TClass::operator())> {};
+
+    template <typename TClass, typename TReturn, typename... TArgs>
+    struct FuncInfo<TReturn(TClass::*)(TArgs...)> : public FuncInfo<TReturn(*)(TArgs...)> {};
+
+    template <typename TClass, typename TReturn, typename... TArgs>
+    struct FuncInfo<TReturn(TClass::*)(TArgs...) const> : public FuncInfo<TReturn(*)(TArgs...)> {};
+
+    template <typename TReturn, typename... TArgs>
+    struct FuncInfo<TReturn(*)(TArgs...)> {};
+
+    template <typename TFunc, typename TReturn, typename... TArgs> requires CFuncReturnsAndIsCallableWithArgs<TFunc, TReturn, TArgs...>
```

Comment:
Can we use these expressions instead of our hand-rolled concepts here?

```
requires std::is_invocable_v<TFunc, TArgs...>
requires std::same_as<TResult, std::invoke_result_t<TFunc, TArgs...>>
```